### PR TITLE
Update GCC to 11.2.0 and binutils to 2.37

### PR DESCRIPTION
**Pull request by rleh**
... and use Github Actions instead of CircleCI.

### CODE REVIEW

**Comment by rleh** [comment]
I get the following error when compiling modm tests:
```
$ cd modm/test/

$ avr-g++ --version
avr-g++ (GCC) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ make compile-arduino-nano_A 
/home/user/.local/bin/lbuild -p ../build/generated-unittest/arduino-nano_A -c config/arduino-nano_A.xml  -C ../build/generated-unittest/arduino-nano_A build --no-log
>> modm: Recomputing device cache...

scons -C ../build/generated-unittest/arduino-nano_A size
scons: Entering directory `/home/user/modm/build/generated-unittest/arduino-nano_A'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Unittest······· unittest_runner.cpp
Compiling C++·· scons-release/unittest_runner.o
as: unrecognized option '-mmcu=avr5'
scons: *** [scons-release/unittest_runner.o] Error 1
scons: building terminated because of errors.
make: *** [Makefile:103: compile-arduino-nano_A] Error 2
```

<details>
<summary>
But the avr-gcc shipped by Fedora 35 also fails:
</summary>

```
$ avr-g++ --version
avr-g++ (Fedora 11.1.0-2.fc35) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ make compile-arduino-nano_A
/home/user/.local/bin/lbuild -p ../build/generated-unittest/arduino-nano_A -c config/arduino-nano_A.xml  -C ../build/generated-unittest/arduino-nano_A build --no-log

scons -C ../build/generated-unittest/arduino-nano_A size
scons: Entering directory `/home/user/modm/build/generated-unittest/arduino-nano_A'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Retrieved `unittest_runner.cpp' from cache
Compiling C++·· scons-release/unittest_runner.o
In file included from modm/src/unittest/controller.hpp:17,
                 from unittest_runner.cpp:1:
modm/src/modm/io/iostream.hpp:25:10: fatal error: inttypes.h: No such file or directory
   25 | #include <inttypes.h>
      |          ^~~~~~~~~~~~
compilation terminated.
scons: *** [scons-release/unittest_runner.o] Error 1
scons: building terminated because of errors.
make: *** [Makefile:103: compile-arduino-nano_A] Error 2
```

</details>

Hmmm 🤔

**Comment by salkinium** [comment]
I had similiar issues using the Homebrew-AVR version, not surprising since this is based on that.

**Comment by salkinium** [comment]
Our avr-libc probably needs to be updated with the v11 version too.

**Comment by rleh** [comment]
Looks like `avr-g++` is calling the wring `as`, because the `avr-as` seems to know `-mmcu=avvr5`:

```
$ /opt/.../avr-gcc/avr-binutils/bin/avr-as --help
Usage: /opt/.../avr-gcc/avr-binutils/bin/avr-as [option...] [asmfile...]
Options:
[...]
AVR Assembler options:
  -mmcu=[avr-name] select microcontroller variant
                   [avr-name] can be:
                   avr1  - classic AVR core without data RAM
                   avr2  - classic AVR core with up to 8K program memory
                   avr25 - classic AVR core with up to 8K program memory
                           plus the MOVW instruction
                   avr3  - classic AVR core with up to 64K program memory
                   avr31 - classic AVR core with up to 128K program memory
                   avr35 - classic AVR core with up to 64K program memory
                           plus the MOVW instruction
                   avr4  - enhanced AVR core with up to 8K program memory
                   avr5  - enhanced AVR core with up to 64K program memory
                   avr51 - enhanced AVR core with up to 128K program memory
                   avr6  - enhanced AVR core with up to 256K program memory
                   avrxmega2 - XMEGA, > 8K, < 64K FLASH, < 64K RAM
                   avrxmega3 - XMEGA, RAM + FLASH < 64K, Flash visible in RAM
                   avrxmega4 - XMEGA, > 64K, <= 128K FLASH, <= 64K RAM
                   avrxmega5 - XMEGA, > 64K, <= 128K FLASH, > 64K RAM
                   avrxmega6 - XMEGA, > 128K, <= 256K FLASH, <= 64K RAM
                   avrxmega7 - XMEGA, > 128K, <= 256K FLASH, > 64K RAM
                   avrtiny   - AVR Tiny core with 16 gp registers
[...]
```

**Comment by rleh** [comment]
Update: I actually works if it the avr-gcc is located in `/opt/avr-gcc`; putting it in `/opt/something/avr-gcc` causes issues...

**Comment by chris-durand** [comment]
> Our avr-libc probably needs to be updated with the v11 version too.

`inttypes.h` is a C header that comes with avr-gcc. That should work, hm ...

EDIT: missed the message from 1 minute earlier that it is some kind of include path issue

**Comment by rleh** [comment]
> `inttypes.h` is a C header that comes with avr-gcc. That should work, hm ...

That is actually a bug in the avr-gcc distributed by the Fedora people. But not our problem currently.

**Comment by rleh** [comment]
Now I get the following error:
```
[...]
Archiving······ scons-release/modm/libmodm.a
sh: line 1: avr-ar: command not found
scons: *** [scons-release/modm/libmodm.a] Error 127
scons: building terminated because of errors.
make: *** [Makefile:103: compile-arduino-nano_A] Error 2
```

Because the executable is actually called `/opt/avr-gcc/avr-gcc/bin/avr-gcc-ar` 😒

**Comment by rleh** [comment]
Update: There also is a `/opt/avr-gcc/avr-binutils/bin/avr-ar`, but I messed up my $PATH...

**Comment by chris-durand** [comment]
> Because the executable is actually called `/opt/avr-gcc/avr-gcc/bin/avr-gcc-ar` :unamused: 

binutils has a `configure` build option to set the name prefix:
```
--program-prefix=PREFIX            prepend PREFIX to installed program names
``` 
Maybe that works to fix the application names.

EDIT: nice that it's solved.

**Comment by rleh** [comment]
Works now!
Review/merge?

File: .github/workflows/build-avr-gcc.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        run: |
+         /opt/avr-gcc/bin/avr-g++ --version
+         /opt/avr-gcc/bin/avr-g++ -mmcu=atmega328p test/main1.cpp
+         /opt/avr-gcc/bin/avr-g++ -mmcu=atmega328p test/main2.cpp
+
+      #- name: Fake toolchain
+      #  run: |
+      #   mkdir -p /opt/avr-gcc/bin
+      #   cp $(which true) /opt/avr-gcc/bin/avr-g++
+
```
Comment (chris-durand) [inline .github/workflows/build-avr-gcc.yml:36]: Should these lines be removed?

File: build.sh [hunk 2]
```diff
@@ -8,55 +8,73 @@ mkdir ${SRC}
...
+VERSION_GCC="11.2.0"
 VERSION_LIBC="2.0.0"
 
 # Get sources
-wget -q   "https://dl.bintray.com/osx-cross/avr-patches/avr-binutils-2.33-size.patch" &
-wget -q   "https://dl.bintray.com/osx-cross/avr-patches/avr-libc-${VERSION_LIBC}-atmega168pb.patch" &
-wget -qO- "https://ftp.gnu.org/gnu/binutils/binutils-2.33.1.tar.bz2" | tar xj --directory ${SRC} &
+wget -q   "https://raw.githubusercontent.com/archlinux/svntogit-community/c3efadcb76f4d8b1a3784015e7c472f59dbfa7de/avr-binutils/repos/community-x86_64/avr-size.patch" &
+wget -q   "https://raw.githubusercontent.com/osx-cross/homebrew-avr/d2e2566b06b90355952ed996707a0a1a24673cd3/Patch/avr-libc-add-mcu-atmega168pb.patch" &
+#wget -q   "https://raw.githubusercontent.com/osx-cross/homebrew-avr/18d50ba2a168a3b90a25c96e4bc4c053df77d7dc/Patch/avr-binutils-elf-bfd-gdb-fix.patch" &
```
Comment (chris-durand) [inline build.sh:19]: Is there a reason why this patch is commented out?

**Review by chris-durand — COMMENTED** [review]
Nice!

File: .github/workflows/build-avr-gcc.yml (hunk 1 above)
Comment (rleh) [reply .github/workflows/build-avr-gcc.yml:36]: I intentially left the lines there, might be useful in the future if CI requires debugging...

Comment (chris-durand) [reply .github/workflows/build-avr-gcc.yml:36]: Then we leave it :+1:

File: build.sh (hunk 2 above)
Comment (rleh) [reply build.sh:19]: The patch only adds `#include <string.h>`, but that does not seem necessary since we don't get an error.

**Review by chris-durand — APPROVED** [review]

**Pull request merged**